### PR TITLE
Add uniquelevels(), fix unique() behaviour

### LINKED ISSUE #357: unique() returns levels, not the CategoricalArray
It's strange that I have realized it only now, but `unique(v::CategoricalArray{S})` returns `Vector{S}`, and not the categorical array of the same type as `v`.
Should it actually preserve the type of `v` to conform to the `Base.unique()` definition?

Otherwise some user code that should work regardless of the vector type (e.g. `Vector{String}` or `CategoricalVector{String}`) produces errors.

### PR DESCRIPTION
**Pull request by alyst**
A draft of a solution for #357, this PR renames `unique(::CategoricalArray)` to `uniquelevels(::CategoricalArray)` as it better reflects what this method does.
As a byproduct, the internal implementation of `uniquelevels()` is cleaned up and optimized a bit (avoid intermediate array sorting, generalize to `AbstractCategoricalArray`).

The PR also re-adds `Base.unique(::CategoricalArray)` with the expected behaviour (returns `CategoricalVector` of unique values).
The new `unique()` is not yet covered by unit tests. I'll do that if the overall approach would be considered ok.

Fixes #357

### CODE REVIEW

File: src/array.jl [hunk 1]
```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
-    order = fill(0, nlevels) # 0 indicates not seen
-    # If we don't track missings, short-circuit even if none has been seen
-    count = S >: Missing ? 0 : 1
-    @inbounds for i in refs
-        if order[i + 1] == 0
-            count += 1
-            order[i + 1] = count
-            count == nlevels && break
+# return unique refs (each value is unique) in the order of appearance in `refs`
+# similar to fallback Base.unique() implementation,
```
Comment (nalimilan) [inline src/array.jl:808]:
```suggestion
# equivalent to fallback Base.unique() implementation,
```

File: src/array.jl [hunk 2]
```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+# but short-circuits once references to all levels are encountered
+function _uniquerefs(A::AbstractCategoricalArray{T}) where T
+    arefs = refs(A)
+    res = similar(arefs, 0)
+    nlevels = length(levels(A))
+    seen = fill(false, nlevels+1) # +1 for 0 (missing ref)
+    maxunique = nlevels + (T >: Missing ? 1 : 0)
+    @inbounds for ref in arefs
+        if !seen[ref + 1]
+            push!(res, ref)
```
Comment (nalimilan) [inline src/array.jl:818]: Instead of pushing here, wouldn't it be more efficient to do `res = (1:maxunique)[seen]` at the end?

File: src/array.jl [hunk 3]
```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
-            count == nlevels && break
+# return unique refs (each value is unique) in the order of appearance in `refs`
+# similar to fallback Base.unique() implementation,
+# but short-circuits once references to all levels are encountered
+function _uniquerefs(A::AbstractCategoricalArray{T}) where T
+    arefs = refs(A)
+    res = similar(arefs, 0)
+    nlevels = length(levels(A))
+    seen = fill(false, nlevels+1) # +1 for 0 (missing ref)
+    maxunique = nlevels + (T >: Missing ? 1 : 0)
```
Comment (nalimilan) [inline src/array.jl:815]:
```suggestion
    maxunique = nlevels + (T >: Missing ? 1 : 0)
    seen = fill(false, maxunique)
```

File: src/array.jl [hunk 4]
```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+function uniquelevels(A::AbstractCategoricalArray{T}) where T
+    alevs = levels(A)
+    T[ref == 0 ? missing : @inbounds alevs[ref]
+      for ref in _uniquerefs(A)]
+end
+
+unique(A::AbstractCategoricalArray{T}) where T =
+    CategoricalVector{T}(_uniquerefs(A), copy(pool(A)))
+
+function unique!(A::CategoricalVector{T}) where T
```
Comment (nalimilan) [inline src/array.jl:842]:
```suggestion
function unique!(A::CategoricalVector)
```

**Review by nalimilan — COMMENTED** [review]
Let's do this then. Unfortunately this is breaking and we released 0.10 recently, so I guess the PR will have to remain unmerged for a while. But better finish it anyway.

File: src/array.jl [hunk 5]
```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+            (length(res) == maxunique) && break
         end
     end
-    S[i == 1 ? missing : levels(pool)[i - 1] for i in sortperm(order) if order[i] != 0]
+    return res
 end
 
 """
-    unique(A::CategoricalArray)
+    uniquelevels(A::AbstractCategoricalArray)
```
[the remark this answers is not in the document]
Comment (alyst) [reply src/array.jl:827]: `unique=` keyword would be a bit obscure, since the levels are unique by definition. It could be something like `byoccurrence=` or `byfirstuse=`.

But I'm not sure this functionality is really needed. It would be nice to keep the returned value of `levels(catarr)` as it is -- independent of the actual stored data and identical to the result of `levels(pool(catarr))` call. And if one needs the original `unique(catarr)`, it would be just `unwrap.(unique(catarr))`. So maybe we just remove `uniquelevels()` and keep only the new `unique()`.

[the remark this answers is not in the document]
Comment (nalimilan) [reply src/array.jl:827]: Yeah, let's not add it for now and see whether `unwrap.(unique(x))` is enough.

**Comment by ablaom** [comment]
This does sound like a good solution.

File: src/array.jl (hunk 2 above)
Comment (alyst) [reply src/array.jl:818]: That would not keep the order of the first appearance of references in `A`.

**Comment by alyst** [comment]
gentle bump

File: src/CategoricalArrays.jl [hunk 6]
```diff
@@ -1,5 +1,6 @@
 module CategoricalArrays
-    export CategoricalPool, CategoricalValue
+
+export CategoricalPool, CategoricalValue
```
Comment (nalimilan) [inline src/CategoricalArrays.jl:3]:
```suggestion
    export CategoricalPool, CategoricalValue
```

File: src/array.jl [hunk 7]
```diff
@@ -854,7 +841,15 @@ end
 Drop levels which do not appear in categorical array `A` (so that they will no longer be
 returned by [`levels`](@ref DataAPI.levels)).
 """
-droplevels!(A::CategoricalArray) = levels!(A, intersect(levels(A), uniquelevels(A)))
+function droplevels!(A::CategoricalArray)
+    # FIXME temporary code until PR#359 is merged
```
Comment (nalimilan) [inline src/array.jl:845]: Just a note so that we don't forget this before merging (x-ref #359).

**Review by nalimilan — COMMENTED** [review]
Looks good! Just waiting for #359 (and then for the next breaking release).

File: src/array.jl (hunk 7 above)
Comment (alyst) [reply src/array.jl]: I've just rebased the PR against the updated `droplevels!()` and squashed the commits, so it should be ready for merging (when it's time for a new minor release).

**Comment by Moelf** [comment]
bump

**Comment by nalimilan** [comment]
This is breaking so it requires tagging a new "major" release. Given the disruption it will create in the ecosystem, I'd like to be sure it's worth it. But so far we don't have lots of breaking changes to make: https://github.com/JuliaData/CategoricalArrays.jl/labels/breaking

**Comment by bkamins** [comment]
The list of breaking things is in https://github.com/JuliaData/CategoricalArrays.jl/labels/breaking. The major decision is with `levels` which is related to `unique`.

There are the following aspects:
* it would be nice to have `levels` and `unique` consistent
* on the other hand there is a risk that a lot of code uses levels exactly to "unwrap" the categorical values into levels. We could emit a warning that this behavior changed, but it is something we never did, and is fragile.
* finally - it would require reviewing the packages depending on CategoricalArrays.jl to update their version bounds.
* in DataFrames.jl all uses of levels and unique are safe to this change

In summary - I think we could go for it, but we should announce on Slack and Discourse that we plan to make this change and wait for a response.

The decision as for `cut` behavior I think is simpler and we could add this behavior while making a breaking release.
